src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/InteropServices/DisableRuntimeMarshalling.cs

```diff
@@ -127,32 +138,237 @@ internal sealed class DisableRuntimeMarshallingAnalyzer : DiagnosticAnalyzer
         public override void Initialize(AnalysisContext context)
         {
             context.EnableConcurrentExecution();
-            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+            // Report diagnostics in generated code to enable this analyzer to catch usages of
+            // delegates with non-blittable parameters that do not have [UnmanagedFunctionPointer] in source-generated interop.
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.ReportDiagnostics);
             context.RegisterCompilationStartAction(context =>
             {
                 if (context.Compilation.TryGetOrCreateTypeByMetadataName(
                     WellKnownTypeNames.SystemRuntimeCompilerServicesDisableRuntimeMarshallingAttribute,
-                    out INamedTypeSymbol? disableRuntimeMarshallingAttribute)
-                    && context.Compilation.Assembly.HasAttribute(disableRuntimeMarshallingAttribute))
+                    out INamedTypeSymbol? disableRuntimeMarshallingAttribute))
                 {
-                    var perCompilationAnalyzer = new PerCompilationAnalyzer(context.Compilation);
-                    perCompilationAnalyzer.RegisterActions(context);
+                    AutoLayoutTypeCache autoLayoutCache = new(context.Compilation);
+                    var hasDisableRuntimeMarshallingAttribute = context.Compilation.Assembly.HasAttribute(disableRuntimeMarshallingAttribute);
+                    if (hasDisableRuntimeMarshallingAttribute)
+                    {
+                        var disabledRuntimeMarshallingAssemblyAnalyzer = new DisabledRuntimeMarshallingAssemblyAnalyzer(context.Compilation, autoLayoutCache);
+                        disabledRuntimeMarshallingAssemblyAnalyzer.RegisterActions(context);
+                    }
+                    var delegateInteropUsageAnalyzer = new DelegateInteropUsageAnalyzer(context.Compilation, autoLayoutCache);
+                    delegateInteropUsageAnalyzer.RegisterActions(context, hasDisableRuntimeMarshallingAttribute);
                 }
             });
         }
 
-        private class PerCompilationAnalyzer
+        private class AutoLayoutTypeCache
         {
-            private readonly INamedTypeSymbol? _unmanagedFunctionPointerAttribute;
             private readonly INamedTypeSymbol? _structLayoutAttribute;
+            private readonly ConcurrentDictionary<ITypeSymbol, bool> _cache = new();
+
+            public AutoLayoutTypeCache(Compilation compilation)
+            {
+                _structLayoutAttribute = compilation.GetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemRuntimeInteropServicesStructLayoutAttribute);
+            }
+
+
+            public bool TypeIsAutoLayoutOrContainsAutoLayout(ITypeSymbol type)
+            {
+                return TypeIsAutoLayoutOrContainsAutoLayout(type, ImmutableHashSet<ITypeSymbol>.Empty.WithComparer(SymbolEqualityComparer.Default));
```

Comment:
I presume this will always return false if `_structLayoutAttribute` is null? If so, can we move the null check out of the local function into the callsite?
```suggestion
                return _structLayoutAttribute is not null && TypeIsAutoLayoutOrContainsAutoLayout(type, ImmutableHashSet<ITypeSymbol>.Empty.WithComparer(SymbolEqualityComparer.Default));
```